compiler/qsc_eval/src/backend.rs

```diff
@@ -264,8 +264,28 @@ impl Backend for SparseSim {
         self.sim.qubit_is_zero(q)
     }
 
-    fn custom_intrinsic(&mut self, name: &str, _arg: Value) -> Option<Result<Value, String>> {
+    fn custom_intrinsic(&mut self, name: &str, arg: Value) -> Option<Result<Value, String>> {
         match name {
+            "GlobalPhase" => {
+                // Apply a global phase to the simulation by doing an Rz to a fresh qubit.
+                // The controls list may be empty, in which case the phase is applied unconditionally.
+                let [ctls_val, theta] = &*arg.unwrap_tuple() else {
+                    panic!("tuple arity for GlobalPhase intrinsic should be 2");
+                };
+                let ctls = ctls_val
+                    .clone()
+                    .unwrap_array()
+                    .iter()
+                    .map(|q| q.clone().unwrap_qubit().0)
+                    .collect::<Vec<_>>();
+                let q = self.sim.allocate();
```

Comment:
Uh oh.... this additional allocation breaks a delicate assumption. Can't yet figure out the concrete ramifications of it though.

https://github.com/microsoft/qsharp/blob/baa771fd4744d52013af268ef4d16dd1ee61e1ea/compiler/qsc_eval/src/backend.rs#L427-L431

Reply:
I guess my question is: does `sim.release()` below decrement the next qubit ID so that the _next_ call to `sim.allocate()` will return the same qubit ID as last time? If the answer is yes, then I think we're in the clear.

Reply:
Yes, good double check. Short answer, you are correct that once a qubit is released it is available to be reused by later calls to allocate. So as you say, we should be safe here.

The longer answer is interesting because it shows how we are relying on some inside knowledge of the sparse simulator to know this short-cut for global phase is actually quite efficient. In the sparse simulator, allocating a new qubit is almost free (some small id maps are upated, but the quantum state itself doesn't have to change at all). Since the qubit starts in the zero state, no bits in the state indices need to change. Then applying Rz with the angle multiplied by -2 causes every state to be multiplied by the desired value that induces the global state without changing the size of the state vector. Finally, releasing the qubit normally requires a collapse of the state, but because the qubit has no entries in the state vector that have a "1" in the corresponding bit, so the release is nearly a no-op (just the map updates). These tricks mean we effectively short-cut to just multiplying the whole state vector by the desired value.